network/network-address/src/encrypted.rs

```diff
@@ -0,0 +1,482 @@
+/// sequence number of the rotation transaction in the `seq_num` field.
+///
+/// ### Key Rotation
+///
+/// The `EncNetworkAddress` struct contains a `root_key_version` field, which
+/// identifies the specific `root_key` used to encrypt/decrypt the
+/// `EncNetworkAddress`.
+///
+/// TODO(philiphayes): expand this section
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct EncNetworkAddress {
+    root_key_version: RootKeyVersion,
+    seq_num: u64,
+    #[serde(with = "serde_bytes")]
+    enc_addr: Vec<u8>,
+}
+
+/// Possible errors when parsing a human-readable [`NetworkAddress`].
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("error encrypting network address")]
+    EncryptError,
+
+    #[error("error decrypting network address")]
+    DecryptError,
+
+    #[error("ciphertext is too small to even contain the auth tag: length: {0}")]
+    CiphertextTooSmall(usize),
+
+    #[error("error deriving account key: {0}")]
+    DeriveAccountKeyError(#[from] HkdfError),
+}
+
+//////////////////////////
+// RawEncNetworkAddress //
+//////////////////////////
+
+impl RawEncNetworkAddress {
+    pub fn new(bytes: Vec<u8>) -> Self {
+        Self(bytes)
+    }
+}
+
+impl Into<Vec<u8>> for RawEncNetworkAddress {
+    fn into(self) -> Vec<u8> {
+        self.0
+    }
+}
+
+impl AsRef<[u8]> for RawEncNetworkAddress {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl TryFrom<&EncNetworkAddress> for RawEncNetworkAddress {
+    type Error = lcs::Error;
+
+    fn try_from(value: &EncNetworkAddress) -> Result<Self, lcs::Error> {
+        let bytes = lcs::to_bytes(value)?;
+        Ok(RawEncNetworkAddress::new(bytes))
+    }
+}
+
+#[cfg(any(test, feature = "fuzzing"))]
+impl Arbitrary for RawEncNetworkAddress {
+    type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        any::<EncNetworkAddress>()
+            .prop_map(|enc_addr| RawEncNetworkAddress::try_from(&enc_addr).unwrap())
+            .boxed()
+    }
+}
+
+///////////////////////
+// EncNetworkAddress //
+///////////////////////
+
+impl EncNetworkAddress {
+    pub fn encrypt(
+        addr: RawNetworkAddress,
+        root_key: &RootKey,
+        root_key_version: RootKeyVersion,
+        account: &AccountAddress,
+        seq_num: u64,
+        addr_idx: u32,
+    ) -> Result<Self, Error> {
+        // unpack the RawNetworkAddress into its base Vec<u8>
+        let mut addr_vec: Vec<u8> = addr.into();
+
+        let account_key = Self::derive_account_key(root_key, account)?;
+        let aead = Aes256Gcm::new(GenericArray::from_slice(&account_key));
+
+        // nonce := seq_num || addr_idx; sizeof(nonce) == 12
+        // ex: seq_num = 0x1234, addr_idx = 0x04
+        //     ==> nonce_slice == &[0, 0, 0, 0, 0, 0, 0x12, 0x34, 0, 0, 0, 0x4]
+        let nonce = ((seq_num as u128) << 32) | (addr_idx as u128);
```

Comment:
I feel like it's less dangerous (and cleaner) to just convert to bytes, then concatenate

Reply:
but muh efficiency : d